guardian/typerighter
# Add client routing

## What does this change?

This PR adds client routing to the application.

To demonstrate the feature, you can turn on the `Show tags page` feature switch in the top right of the client-side application, which will add a navbar to the header, allowing you to navigate between pages.

![image](https://github.com/guardian/typerighter/assets/34686302/dca38c65-b885-4692-b081-e9943d720c55)

Any top-level path will now be handed to the client-side React application, unless it is specifically identified in the Play `routes` file (e.g. `/oauthCallback`).

Our other existing routes are now moved to `/api/[path here]`.

I've also added a charming 404 page in the unlikely event that someone types in a non-existent top-level path.

![image](https://github.com/guardian/typerighter/assets/34686302/31ea36f4-017e-445c-b206-6bb946d8b4ca)
 
## How to test

1. Run typerighter locally according to the instructions in the readme
2. Try standard rule creation and publication steps. Do you get any server errors in your network tab? This may indicate that an api path hasn't been updated.
3. Try navigating directly to the url `/tags` via your browser url bar. Does it work?
4. Activate the 'Show tag page' feature switch and try clicking the nav links. Do they take you to expected page?
5. Try navigating directly to the url `/oauthCallback` via your browser. Is it being handled by the backend? (It should be!)
6. Try navigating directly to the url `api/rules` via your browser. Do you see some JSON provided by the backend?
7. Try navigating to a top-level fake path, e.g. `/cheese`. Do you see the 404 page?
